sheets/api.py

```diff
@@ -253,31 +261,111 @@ def batch_update_sheet_cells(self, sheet_id, request_objects):
-                    "type": GOOGLE_API_NOTIFICATION_TYPE,
-                },
-            )
-            .execute()
+def request_file_watch(file_id, channel_id, expiration=None, credentials=None):
+    """
+    Sends a request to the Google API to watch for changes in a given file. If successful, this
+    app will receive requests from Google when changes are made to the file.
+    Ref: https://developers.google.com/drive/api/v3/reference/files/watch
+
+    Args:
+        file_id (str): The id of the file in Google Drive (can be determined from the URL)
+        channel_id (str): Arbitrary string to identify the file watch being set up. This will
+            be included in the header of every request Google sends to the app.
+        expiration (datetime.datetime or None): The datetime that this file watch should expire.
+            Defaults to 1 hour, and cannot exceed 24 hours.
+        credentials (google.oauth2.credentials.Credentials or None): Credentials to be used by the
+            Google Drive client
+
+    Returns:
+        dict: The Google file watch API response
+    """
+    credentials = credentials or get_credentials()
+    # Including cache_discovery=False to suppress a noisy warning
+    drive_service = build("drive", "v3", credentials=credentials, cache_discovery=False)
+    extra_body_params = {}
+    if expiration:
+        extra_body_params["expiration"] = format_datetime_for_google_timestamp(
+            expiration
+        )
+    return (
+        drive_service.files()
+        .watch(
+            fileId=file_id,
+            supportsTeamDrives=True,
+            body={
+                "id": channel_id,
+                "resourceId": file_id,
+                "address": urljoin(
+                    settings.SITE_BASE_URL,
+                    reverse("handle-coupon-request-sheet-update"),
+                ),
+                "payload": True,
+                "kind": GOOGLE_API_FILE_WATCH_KIND,
+                "type": GOOGLE_API_NOTIFICATION_TYPE,
+                "expiration": "failfailfail"
+                # **extra_body_params,
+            },
         )
+        .execute()
+    )
+
+
+def renew_coupon_request_file_watch():
+    """
+    Creates or renews a file watch on the coupon request spreadsheet depending on the existence
+    of other file watches and when they expire.
+
+    Returns:
+        (GoogleFileWatch, bool): The GoogleFileWatch object paired with a flag indicating
+            whether or not it was newly created during execution.
+    """
+    now = now_in_utc()
+    min_fresh_expiration_date = now + datetime.timedelta(
+        minutes=settings.DRIVE_WEBHOOK_RENEWAL_PERIOD_MINUTES
+    )
+    file_watches = GoogleFileWatch.objects.filter(
+        file_id=settings.COUPON_REQUEST_SHEET_ID
+    ).order_by("-expiration_date")
+    latest_watch = file_watches.first()
+    if latest_watch and latest_watch.expiration_date > min_fresh_expiration_date:
+        return latest_watch, False
```

Comment:
This function should be run at least every `settings.DRIVE_WEBHOOK_RENEWAL_PERIOD_MINUTES` minutes?

Reply:
The way I configured the settings, `DRIVE_WEBHOOK_EXPIRATION_MINUTES` is the total time to live for the webhook, and `DRIVE_WEBHOOK_RENEWAL_PERIOD_MINUTES` is the "cushion", i.e.: we will try to renew the webhook this many minutes before the expiration arrives. Do you think it makes more sense just to change `DRIVE_WEBHOOK_RENEWAL_PERIOD_MINUTES` so that it represents the total time we will wait to renew? We could assert that it's less than the expiration period. If so, that's an easy change. Let me know what you think

Reply:
Either way is probably fine, I was concerned that the task run might get slightly out of sync where `latest_watch.expiration_date > min_fresh_expiration_date` might be True but just barely, and then not run again til long after it expired.  Looking at the code again though I don't think this would happen.